### PR TITLE
emacsPackages.org-roam: init at 0.1.2

### PR DESCRIPTION
**Pull request by rasendubi**
###### Motivation for this change
org-roam seems to be a nice package, but it's not packaged into melpa/elpa

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

I'm running it now and it works fine.

### CODE REVIEW

**Comment by adisbladis** [comment]
Could you try to get upstream to upload the package to elpa/melpa?

**Comment by rasendubi** [comment]
Sure. Asked for that in https://github.com/jethrokuan/org-roam/issues/171

**Comment by jethrokuan** [comment]
I'd like to upload to MELPA when things become more stable, which is when I decide it's version `1.0.0`.

**Comment by rasendubi** [comment]
A melpa release won't happen any time soon. The question is, do we want org-roam in nixpkgs now, or should I stash it in my overlay?

I would go for nixpkgs as org-roam is quite an active project and gets excellent traction ([see pulse](https://github.com/jethrokuan/org-roam/pulse/monthly)). I've been using 0.1.2 for two weeks now, and I find it quite stable and useful.

@adisbladis, what do you think?

**Comment by jethrokuan** [comment]
> A melpa release won't happen any time soon. 

Within the next week I think!

**Comment by rasendubi** [comment]
oh wow. that's great news! I think we should wait then :)

**Comment by adisbladis** [comment]
Great! I was just going to say "let's just add it", but if a proper MELPA release is coming soon then let's wait for it :)

File: pkgs/applications/editors/emacs-modes/manual-packages.nix [hunk 1]
```diff
@@ -110,6 +110,27 @@
...
+      rev = "v${version}";
+      sha256 = "1p8bhj09s0iyb3fcjibsl1p61wiac17sn1w7hwm1wqrcydj8h8hx";
+    };
+    recipe = pkgs.writeText "recipe" ''
+      (org-roam
+       :repo "jethrokuan/org-roam"
+       :fetcher github
+       :files ("*.el"))
+    '';
+    packageRequires = [ dash f s async ];
```
Comment (jethrokuan) [inline pkgs/applications/editors/emacs-modes/manual-packages.nix]: for when 1.0.0 lands:

```suggestion
    packageRequires = [ dash f s emacsql emacsql-sqlite ];
```

Comment (adisbladis) [reply pkgs/applications/editors/emacs-modes/manual-packages.nix]: @jethrokuan If the package is pulled from melpa all those bit's are automagic.

**Comment by adisbladis** [comment]
I'm closing this for now since the package will arrive in Nixpkgs through the usual MELPA workflow.

**Comment by JeffreyBenjaminBrown** [comment]
Can we now install org-roam from our NixOS configuration?

I've tried the obvious way, by adding `org-roam` to my `epkgs.melpaStablePackages` definitions, and I get the following error when I try to build:
```
error: undefined variable 'org-roam' at /etc/nixos/emacs.nix:16:7
```

Two of you have made public your NixOS configurations; I cloned them and grepped (case insensitive) for `roam` and found nothing.

Here's [the commit where I made the change](https://github.com/JeffreyBenjaminBrown/nixos-experiments/commit/8c652d41f7eb04234adbe393502e2b113cff712a), and here's the [complete file including that change](https://github.com/JeffreyBenjaminBrown/nixos-experiments/blob/org-roam/emacs.nix).

**Comment by rasendubi** [comment]
Hey @JeffreyBenjaminBrown!

It's in `melpaPackages`, not `melpaStablePackages`. Move org-roam just a couple of lines below under `with melpaPackages`—that should help.

**Comment by JeffreyBenjaminBrown** [comment]
I tried putting it under melpa instead of melpaStable. I even tried under elpa. They all give me the same error, except with a different line number.

It's as if Nix is looking at an old Melpa repo when it tries to build my config. Your config includes `company`, which I didn't know about and looks cool, so I tried adding that to my `melpaPackages` declarations and it installs fine. But company-roam and org-roam generate errors when I do likewise with them.

So I tried adding the overlay you mentioned. First time overlaying, and I don't know if I added it the right way. [How I added the overlay](https://github.com/JeffreyBenjaminBrown/nixos-experiments/commit/ab89f0ab43b97cc6ab0948bee2de55b1358ebbd0) didn't generate an error but neither did it solve this problem.

I tried to understand your .org file that turns into your .nix files. Oh my god, I haven't felt so stupid in years; I don't even know what I would have to study to understand it. When I run `org-babel-tangle` it actually deletes the three lines containing the word `roam` from the `.nix` files in your repo, but I tried studying those files after restoring them to their original state and can't make heads or tails of it.

**Comment by adisbladis** [comment]
`{ pkgs ? import <nixpkgs> {} }: {}` is syntax for "create a function that takes an attrset where `pkgs` _defaults_   to `import <nixpkgs> {}`".

The thing is you are _explicitly passing_ `pkgs` here https://github.com/JeffreyBenjaminBrown/nixos-experiments/blob/master/packages.nix#L5 which makes the entire overlay addition moot.

What you should be doing is adding the overlay to the NixOS option `nixpkgs.overlay` as shown in the overlay README: https://github.com/nix-community/emacs-overlay#latest-master-each-rebuild.

**Comment by JeffreyBenjaminBrown** [comment]
Thanks! I thought that might be happening, but it was the only thing I tried that compiled.

I had looked at the section of the emacs-overlay README you mention, and also the NixOS wiki article on overlays, and was totally baffled. I just started a new [thread on the NixOS Discourse to tackle that problem](https://discourse.nixos.org/t/my-first-overlay-for-emacs/7570).

**Pull request closed without merging**